File: 1074 - Number of Submatrices That Sum to Target/solution.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& A, int target) {
        int res = 0, m = A.size(), n = A[0].size();
        for(int i = 0; i < m; i++)
            for(int j = 1; j < n; j++)
                A[i][j] += A[i][j-1];
        unordered_map<int, int> um;
        for(int i = 0; i < n; i++){
            for(int j = i; j < n; j++){
                um = {{0, 1}};
                int total = 0;
                for(int k = 0; k < m; k++){
                    total += A[k][j]-(i?A[k][i-1]:0);
                    if(um.count(total-target))
                        res += um[total-target];
                    um[total]++;
                }
            }
        }
        return res;
    }
};

File: 1074 - Number of Submatrices That Sum to Target/solution.py
from itertools import combinations_with_replacement
class Solution:
    def numSubmatrixSumTarget(self, A: List[List[int]], target: int) -> int:
        res, m, n = 0, len(A), len(A[0])
        for i in range(m):
            for j in range(1, n):
                A[i][j] += A[i][j-1]
        for i, j in combinations_with_replacement(range(n), 2):
            total, mp = 0, {0: 1}
            for k in range(m):
                total += A[k][j]-(A[k][i-1] if i else 0)
                res += mp.get(total-target, 0)
                mp[total] = mp.get(total, 0)+1
        return res